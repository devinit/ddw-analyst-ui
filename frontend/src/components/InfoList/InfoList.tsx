import * as React from 'react';
import { OverlayTrigger, Popover, Table } from 'react-bootstrap';
import { List, Map } from 'immutable';
import { ReactNode, ReactElement } from 'react';

export interface InfoItem {
  caption: string;
  info: string;
}
export type InfoMap = Map<keyof InfoItem, InfoItem[keyof InfoItem]>;
export type InfoListItems = List<InfoMap>;
interface InfoListProps {
  list: InfoListItems;
  className?: string;
}

export class InfoList extends React.Component<InfoListProps> {
  render(): ReactElement {
    return (
      <div>
        <Table size="sm" className={`${this.props.className || ''}`}>
          <tbody>
            {this.props.list.map((item, index) => (
              <tr key={index}>
                <td colSpan={item.get('info') ? 1 : 2}>{item.get('caption')}</td>
                {this.renderRow(item)}
              </tr>
            ))}
          </tbody>
        </Table>
      </div>
    );
  }

  private renderRow(item: InfoMap): ReactNode {
    const info = item.get('info');
    if (info) {
      return (
        <td className="text-right">
          <OverlayTrigger placement="bottom" overlay={this.renderPopOver(info)}>
            <i className="material-icons" data-testid="info-trigger">
              info
            </i>
          </OverlayTrigger>
        </td>
      );
    }
  }

  private renderPopOver(info: string): ReactElement {
    return (
      <Popover id="popover-basic" data-testid="info-list-info">
        {info}
      </Popover>
    );
  }
}
