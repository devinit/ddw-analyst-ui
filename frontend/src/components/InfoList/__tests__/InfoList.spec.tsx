/**
 * @jest-environment jsdom
 */
import * as React from 'react';
import { fireEvent, render, waitForElement } from 'react-testing-library';
import { InfoList, InfoListItems, InfoMap } from '../InfoList';
import { List, Map } from 'immutable';

let list: InfoListItems;

beforeEach(() => {
  list = List([ Map({ caption: 'Item 1', info: 'My First Item' }) as InfoMap ]);
});

test('renders correctly with the default props', () => {
  const { container } = render(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();
});

test('renders all list items', async () => {
  list = list.push(Map({ caption: 'Item 2', info: 'My Second Item' }) as InfoMap);
  expect(list.count()).toEqual(2);
  const { container } = render(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();
});

test('updates when the list changes', () => {
  const { container, rerender } = render(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();

  list = list.push(Map({ caption: 'Item 2', info: 'My Second Item' }) as InfoMap);
  rerender(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();
});

test('renders the info on hover', async () => {
  const { getByTestId } = render(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  fireEvent.mouseEnter(getByTestId('info-trigger'));
  const popOver = await waitForElement(() => getByTestId('info-list-info'));

  expect(popOver).toMatchSnapshot();
});

test('does not render the info trigger when no info is provided', async () => {
  list = list.set(0, Map({ caption: 'Item 1', info: '' }) as InfoMap);
  const { container } = render(<InfoList list={ list } limit={ 10 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();
});

test('renders the pagination when the list exceeds the page limit', async () => {
  for (let count = 1; count < 12; count++) {
    list = list.push(Map({ caption: `Item ${count + 1}`, info: `Item Number ${ count + 1}` }) as InfoMap);
  }
  const { container, getByTestId } = render(<InfoList list={ list } limit={ 5 } offset={ 0 }/>);

  expect(container).toMatchSnapshot();

  fireEvent.click(getByTestId('info-pagination-next'));
  expect(container).toMatchSnapshot();

  fireEvent.click(getByTestId('info-pagination-prev'));
  expect(container).toMatchSnapshot();

  fireEvent.click(getByTestId('info-pagination-last'));
  expect(container).toMatchSnapshot();

  fireEvent.click(getByTestId('info-pagination-first'));
  expect(container).toMatchSnapshot();
});
