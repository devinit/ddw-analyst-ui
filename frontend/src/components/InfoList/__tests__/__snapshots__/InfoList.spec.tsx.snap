// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render the info trigger when no info is provided 1`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="2"
          >
            Item 1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders all list items 1`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 2
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders correctly with the default props 1`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders the info on hover 1`] = `
<div
  class="fade show popover bs-popover-bottom"
  data-testid="info-list-info"
  id="popover-basic"
  role="tooltip"
  style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
  x-placement="bottom"
>
  <div
    class="arrow"
  />
  <div
    class="popover-body"
  >
    My First Item
  </div>
</div>
`;

exports[`renders the pagination when the list exceeds the page limit 1`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 2
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 3
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 4
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 5
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="row"
    >
      <div
        class="col-lg-6"
      >
        Showing 
        1
         to 
        5
         of 
        12
      </div>
      <div
        class="col-lg-6"
      >
        <ul
          class="float-right pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-first"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  first_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-prev"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_left
                </i>
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-next"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_right
                </i>
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-last"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  last_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the pagination when the list exceeds the page limit 2`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 6
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 7
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 8
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 9
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 10
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="row"
    >
      <div
        class="col-lg-6"
      >
        Showing 
        6
         to 
        10
         of 
        12
      </div>
      <div
        class="col-lg-6"
      >
        <ul
          class="float-right pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-first"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  first_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-prev"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_left
                </i>
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-next"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_right
                </i>
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-last"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  last_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the pagination when the list exceeds the page limit 3`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 2
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 3
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 4
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 5
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="row"
    >
      <div
        class="col-lg-6"
      >
        Showing 
        1
         to 
        5
         of 
        12
      </div>
      <div
        class="col-lg-6"
      >
        <ul
          class="float-right pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-first"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  first_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-prev"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_left
                </i>
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-next"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_right
                </i>
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-last"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  last_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the pagination when the list exceeds the page limit 4`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 11
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 12
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="row"
    >
      <div
        class="col-lg-6"
      >
        Showing 
        11
         to 
        12
         of 
        12
      </div>
      <div
        class="col-lg-6"
      >
        <ul
          class="float-right pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-first"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  first_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-prev"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_left
                </i>
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-next"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_right
                </i>
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-last"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  last_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the pagination when the list exceeds the page limit 5`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 2
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 3
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 4
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 5
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="row"
    >
      <div
        class="col-lg-6"
      >
        Showing 
        1
         to 
        5
         of 
        12
      </div>
      <div
        class="col-lg-6"
      >
        <ul
          class="float-right pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-first"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  first_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-prev"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_left
                </i>
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-next"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  chevron_right
                </i>
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="info-pagination-last"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                <i
                  class="material-icons"
                >
                  last_page
                </i>
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`updates when the list changes 1`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`updates when the list changes 2`] = `
<div>
  <div>
    <table
      class="table table-sm"
    >
      <tbody>
        <tr>
          <td
            colspan="1"
          >
            Item 1
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Item 2
          </td>
          <td
            class="text-right"
          >
            <i
              class="material-icons"
              data-testid="info-trigger"
            >
              info
            </i>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
